### ISSUE TITLE
macOS: balena-cli can't be opened because Apple cannot check it for malicious software

### ISSUE THREAD

**Issue by pdcastro** [balena-io/balena-cli#1479]
- **balena CLI version:** any (e.g. 12.27.3)
- **Operating system version:** macOS 10.15 (Catalina) and 11.0 (Big Sur)
- **32/64 bit OS and processor:** any
- **Install method:** executable (native) installer and standalone zip

Apple has introduced the requirement of "[software notarization](https://developer.apple.com/documentation/security/notarizing_your_app_before_distribution)" for macOS, but the balena CLI installers have not yet been updated to include this step. As a result, macOS will raise warning pop-ups before the CLI can be installed or executed. Until the CLI installers are notarized, please follow the workaround steps below.

<img width="532" alt="native-error" src="https://user-images.githubusercontent.com/15091591/67145203-5570da00-f277-11e9-8823-176e5fd29012.png">

**Workaround steps for the graphical macOS installer (.pkg)**

1. After downloading the executable (native) installer (`.pkg` file), [right-click](https://support.apple.com/en-gb/HT207700) (2-finger tap or CTRL-click) the file and select "Open" in the pop-up menu:

<img width="559" alt="native-open-menu" src="https://user-images.githubusercontent.com/15091591/67145315-78e85480-f278-11e9-8fe2-5e0f8b281b46.png">

&nbsp;
2. A warning pop-up will be raised, including an **Open** button. **Click the Open button:**

<img width="532" alt="native-open-button" src="https://user-images.githubusercontent.com/15091591/67145465-c618f600-f279-11e9-898e-dec441f92df6.png">

3. The executable (native) installer will open: proceed with the installation, [as usual](https://github.com/balena-io/balena-cli/blob/master/INSTALL.md).

&nbsp;  
If you are using the standalone zip package, check the alternative instructions in the next comment.

**Comment by pdcastro** [comment]
Connects to #1376 (CLI package for Mac needs to be notarized)

**Comment by pdcastro** [comment]
Hints for the implementation:
* https://github.com/mitchellh/gon
* https://www.flowdock.com/app/rulemotion/r-etcherprv/threads/VvGI_yrIc-5f9ScEYeovzVWwrkO

**Comment by jesperbagge** [comment]
I've run into additional issues with the standalone package (v11.19.1) on macOS Catalina that can't be resolved with this workaround. macOS also complains on the file **bindings.node**, which I can't get around by right-clicking.

**Comment by pdcastro** [comment]
@jesperbagge, thanks for reporting this. I have added "step 4" to the instructions, showing how to get around the issue with **bindings.node** and other `*.node` libraries. Apologies for this frustrating experience. The extra steps may only be required for the standalone zip file (not completely clear), so you may find the executable installer preferable until we get Apple notarization sorted (soon).

**Comment by shawaj** [comment]
Just an FYI - I still get this on Mac OS Big Sur...

![image](https://user-images.githubusercontent.com/3359418/102032173-0574a100-3db0-11eb-9b78-1f02006f44df.png)

**Comment by uvaaland** [comment]
I am having the same problem on Mac OS Big Sur, even with the "Open Anyway" option in Security & Privacy.

**Comment by shawaj** [comment]
@uvaaland if you go to it in Finder, right click and then choose open from the drop-down and you should be able to proceed.

**Comment by uvaaland** [comment]
@shawaj I tried that initially, but I get the same message that you posted above:

<img width="240" alt="issue" src="https://user-images.githubusercontent.com/13985236/104121755-f77b5700-5340-11eb-89e1-71f5fc199ec9.png">

**Comment by shawaj** [comment]
@uvaaland when I go to finder, right click, select Open I get the following:

![image](https://user-images.githubusercontent.com/3359418/104168634-3fa88100-53f6-11eb-82a0-d83b2bce4bc8.png)

I guess you opened it by double clicking on it? Not by right clicking and selecting Open. When I do that, I get the same as you:

![image](https://user-images.githubusercontent.com/3359418/104168730-6666b780-53f6-11eb-8bd9-761fdd111d97.png)

**Comment by uvaaland** [comment]
@shawaj No, I am opening it by right-clicking and selecting Open. That is, whether I try to open it by double-clicking or right-click and Open, I get the same message that I posted above.

**Comment by shawaj** [comment]
@uvaaland weird. Maybe something in your mac settings? Because for me it offers a different option depending on how I open it

**Comment by pdcastro** [comment]
This issue was resolved in CLI release v12.44.4 or later, at least for the graphical macOS installer (`.pkg` file extension). See [installation instructions](https://github.com/balena-io/balena-cli/blob/master/INSTALL-MAC.md).

Regarding the standalone zip package, it is still not notarized but I have learned about a simpler workaround that avoids the convoluted series of popups. It is described in a new issue, #2244.
